### ISSUE TITLE
KarmaService Improvement

### ISSUE THREAD

**Issue by IanKeen** [iosdevelopershq/camille#4]
Currently KaramBot supports `++` and `--` of users however it's still behind Camille in terms of functionality, we need to add:
- [x] Leaderboard, able to be truncated to a specific number: i.e. `top 10`

In addition it would be nice to add some new stuff as well:
- [ ] when you `--` a user I think it might be worthwhile requiring a reason

**Comment by rosslebeau** [comment]
Working on top 10 now!

**Comment by woolsweater** [comment]
Do you want the bot to do something with the `--` reason, or do you just want it to not affect parsing of the message?

**Comment by IanKeen** [comment]
So I think it would be cool if the flow was something like:

1: <userA> @fooUser++
2: bot DMs the user asking why they want to take away karma.
-- we should provide a timeout too, so if they dont respond soon enough nothing happens (maybe a cheeky message in the channel from the bot? :P )
3: user responds with reason
4: bot decreases karma and shows message and reason in the channel

Thoughts?

**Comment by woolsweater** [comment]
I'm not sure why the switch to DM to collect the reason. Why not require it up front? I.e., the message must be `@badUser-- Don't insult other people's mothers` and a bare `@badUser--` would produce no result.

Or a missing reason could cause a prompt in an DM.

**Comment by IanKeen** [comment]
so the reason I suggested DM is because of the way ++ and -- now works. I can say something like:

`<Ian> hey thanks for the help @woolsweater++ and @rosslebeau++ couldn't have done it without you!`

the same is possible with `--` though I'm not sure people would use it that way

**Comment by IanKeen** [comment]
from #bot-laboratory:

after a user does a `--` a better option might be to have the bot DM the user with a series of preset reasons (as buttons) 

we could also possibly have certain presets alert admins (such as abusive behaviour) etc..

**Comment by rosslebeau** [comment]
Just to note: top 10 is done and merged

**Comment by IanKeen** [comment]
Quick thought, would be cool if Camille added a :balloon: reaction to the karma message if its someones first karma ++ 😄

**Comment by woolsweater** [comment]
Nothing wrong with a balloon!

I was mulling over the proposal that everyone start at 1 karma -- just change the return from a failed score lookup to 1 instead of 0. 

This would be a kind of "we assume everyone is good people" gesture, and I thought it would align pretty well with Ryan's ["make karma more friendly"](https://github.com/aranasaurus/camille-js/issues/1) issue on the original Camille.

**Comment by IanKeen** [comment]
Ohh I hadn't seen that issue, yeah I think either of those changes would make karma less intimidating for new comers which is 👍 
It may require us to alter how karma is stored (if we are doing cumulative tracking) but this also opens it up for us to find other cool stats about karma. i.e.: who gives out the most, who gave user X most of their karma etc..

Tagging @aranasaurus in this also :D

**Comment by aranasaurus** [comment]
Yeah, I'm hoping to get some time to look at this soon. I want to add a bunch of stuff to how the KarmaService stores data in order to do a lot of the things I wanted to do with the original but was simply unable to 'cause: CoffeeScript :P
